core/dbt/task/compare.py

```diff
@@ -0,0 +1,82 @@
+from __future__ import print_function
+
+from dbt.adapters.factory import get_adapter
+from dbt.logger import GLOBAL_LOGGER as logger
+from dbt.runner import RunManager
+from dbt.node_types import NodeType
+from dbt.node_runners import ModelRunner
+from dbt.utils import is_enabled as check_is_enabled
+
+import dbt.ui.printer
+from dbt.task.base_task import BaseTask
+
+
+class CompareTask(BaseTask):
+    def _get_manifest(self):
+        compiler = dbt.compilation.Compiler(self.config)
+        compiler.initialize()
+
+        all_projects = compiler.get_all_projects()
+
+        manifest = dbt.loader.GraphLoader.load_all(self.config, all_projects)
+        return manifest
+
+    def run(self):
+
+        logger.info(
+            "dbt compare: Comparing local codebase to catalog in data warehouse."
+        )
+
+        manifest = self._get_manifest()
+
+        # Look up all of the relations dbt knows about
+        model_relations = set()
+        for node in manifest.nodes.values():
+            node = node.to_dict()
+            is_refable = node["resource_type"] in NodeType.refable()
+            is_enabled = check_is_enabled(node)
+            is_ephemeral = node["config"]["materialized"] == "ephemeral"
+            if is_refable and is_enabled and not is_ephemeral:
+                rel = (node["schema"].lower(), node["alias"].lower())
+                model_relations.add(rel)
+
+        # Look up all of the relations in the DB
+        adapter = get_adapter(self.config)
+        results = adapter.get_catalog(manifest)
```

Comment:
I took a deeper look at this and don't think we actually want to use `get_catalog` here. That method returns the full catalog, which includes all of the relations in the specified schemas, PLUS all of the columns in those relations. Instead, you should be able to just use `list_relations`, eg:

```
relations = []
for schema in manifest.get_used_schemas():
  relations.extend(adapter.list_relations(schema))
```

This will make `results` a list of [Relation](https://docs.getdbt.com/reference#relation) objects. Relations are great because they know how to quote themselves (see below):

```
ipdb> my_relation.schema
'demo_schema'
ipdb> my_relation.identifier
'my_table'
ipdb> my_relation
<BaseRelation "demo_schema"."my_table">
ipdb> str(my_relation)
'"demo_schema"."my_table"'
```

I'd do two things with these relations:
1) convert them into a tuple (as you already do below)
2) add an entry to a dictionary which maps the tuple onto the original relation

Eg:
```
database_relations = set()
database_relations_map = dict()
for relation in relations:
    relation_id = (relation.schema.lower(), relation.identifier.lower())
    database_relations_map[relation_id] = relation
    database_relations.add(relation_id)
```

If you do this, then you can do a set difference below to calculate the extraneous db relations, then _look up the original relation object_. This will help with effective quoting, again: see below